### PR TITLE
prof: provide extra mapping info in pprof dumps

### PR DESCRIPTION
**Pull request by teskje**
This PR extends the pprof memory profiles we produce by a set of `executableInfo` comments. The information in these comments is required for offline symbolization to work correctly.

To retrieve the `executableInfo` information, the `mz-build-id` crate is renamed to `mz-proc` and extended to not just retrieve build IDs but also memory offsets and paths of all objects loaded into the process space. This information is sufficient to also generate the pprof `mappings`, which means that we don't have to parse those from the jemalloc output anymore.

Note that we are now including fewer mappings in the pprof dump that we did before this change: Only mappings of files are included, anonymous mappings are skipped. That is fine, since for anonymous mappings we don't have symbol information anyway, so including them into the pprof output wouldn't give us anything.

### Motivation

  * This PR fixes a recognized bug.

Fixes https://github.com/MaterializeInc/materialize/issues/22451

### Tips for reviewer

Look at the commits separately. The 3rd one contains the meat of the changes. The 1st one renames `mz-build-id` to `mz-proc` and moves the code within a bit, the 2nd simplifies `string_table` management in pprof writing.

See also [this Slack thread](https://materializeinc.slack.com/archives/C049UQ3RSF4/p1697449947110039) for context.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by teskje** [comment]
I tested this on staging and the results look good!
Screenshot 2023-10-19 at 13 28 04

File: src/prof/src/lib.rs [hunk 1]
```diff
@@ -318,6 +289,34 @@ mz_fg_version: 1
...
+        }
+    }
+
+    fn finish(self) -> Vec<String> {
+        let mut vec: Vec<_> = self.0.into_iter().collect();
+        vec.sort_by_key(|(_, idx)| *idx);
+        vec.into_iter().map(|(s, _)| s).collect()
+    }
+}
+
```
Comment (umanwizard) [inline src/prof/src/lib.rs:342]: Thanks, this `StringTable` abstraction simplifies the code a lot.

File: src/proc/src/linux.rs [hunk 2]
```diff
@@ -89,127 +92,135 @@ extern "C" fn iterate_cb(info: *mut dl_phdr_info, _size: size_t, data: *mut c_vo
...
+                offset = offset
+                    + std::mem::size_of::<NoteHeader>()
+                    + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_namesz))
+                    + align_up::<ELF_NOTE_STRING_ALIGN>(usize::cast_from(nh.n_descsz));
             }
         }
     }
-    0
+
+    let objects = state.result.as_mut().expect("we return early on errors");
```
Comment (umanwizard) [inline src/proc/src/linux.rs:215]: What will happen if we panic here? Is it okay in Rust to panic in a function that's called as a callback by some C function?

File: src/prof/src/lib.rs [hunk 3]
```diff
@@ -223,6 +226,26 @@ mz_fg_version: 1
...
+            // field for the file offset. So we instead encode additional information about
+            // mappings in magic comments. There must be exactly one comment for each mapping.
+
+            // Take a shortcut and assume the ELF type is always `ET_DYN`. This is true for shared
+            // libraries and for position-independent executable, so it should always be true for
+            // any mappings we have.
+            // Getting the actual information is annoying. It's in the ELF header (the `e_type`
+            // field), but there is no guarantee that the full ELF header gets mapped, so we might
+            // not be able to find it in memory. We could try to load it from disk instead, but
+            // then we'd have to worry about blocking disk I/O.
```
Comment (umanwizard) [inline src/prof/src/lib.rs:241]: Note that if it is actually mapped, I don't think the I/O will block, as the page will be in the filesystem cache, if I understand correctly how Linux works.

(Btw, I'm not arguing that we should actually try to do file i/o here... this hack is fine)

**Review by umanwizard — APPROVED** [review]
This is great, thanks!

File: src/proc/src/linux.rs (hunk 2 above)
Comment (teskje) [reply src/proc/src/linux.rs:215]: It's a good question! I searched around a bit and this seems like a good summary: https://users.rust-lang.org/t/passing-callbacks-to-c-panic/91080/20

TLDR: Since Rust 1.71 `extern "C"` functions always abort on panic. But in Mz we are configured to abort on panic anyway, so would have done that even in earlier Rust versions.

**Comment by teskje** [comment]
TFTR!
